pkg/ebpf/c/tracee.bpf.c

```diff
@@ -6010,9 +6083,22 @@ int cgroup_rmdir_signal(struct bpf_raw_tracepoint_args *ctx)
     save_to_submit_buf(&signal->args_buf, &cgroup_id, sizeof(u64), 0);
     save_str_to_buf(&signal->args_buf, path, 1);
     save_to_submit_buf(&signal->args_buf, &hierarchy_id, sizeof(u32), 2);
-    signal_perf_submit(ctx, signal, CGROUP_RMDIR);
 
-    return 0;
+    return signal_perf_submit(ctx, signal, CGROUP_RMDIR);
 }
 
-// clang-format on
+//
+// NOTE: The sched_process_{fork,exec,exit} control events are submitted by the
```

Comment:
Ok so let's get to the (non) separation choice you made.

I have chosen to write the initial control plane programs independently of the previous programs in a conscious choice so that we can have the control plane be a seamless background process independent of the event mechanisms of tracee, only for signalling and updating relevant infrastructure ebpf maps. Now I recognize the overhead this causes (and of course for the cgroup programs this is less of an issue), and I assume we can maybe do a shared map, if we know for sure that the control plane programs runs before the event programs.
This is why I attach the control plane first, though i'm not certain eBPF actually synchronizes program order like that, please correct me if i'm wrong, because if I am, separation is not warranted at all.

So we should be consistent about this, either we have it always separate, with a possible dependency of the event program on the control plane program (because the control plane by it's nature should be independent of any event program) for initial argument saving, shared by a map to avoid the double parsing.
OR, we can convert the cgroup programs to what you did, and have them submit a signal first, and then after event tracing validation that we do in standard programs, the event submission.

Reply:
In my mind I thought that both perbuffers (and even some other events other than ProcessSchedule, if needed) could update the tree but eventually the control buffer would be faster due to less queueing. 

Of course I still haven't coded that and wanted feedback first.

Reply:
Not sure how this relates to what I've commented by maybe I didn't phrase it clearly enough.

I'm discussing here how you've opted to do signal submission and event submission in the same eBPF program, instead of using separate ones like I did in the cgroup ones. So there's a lack of consistency in that.
So the question is whether we should be consistent to your decision (and merge the cgroup programs), or be consistent to my decision and split the program. I think the sched programs aren't the most taxing ones we have (that honor goes to the syscalls and security_file_open) so I don't think the duplication would be too bad.

Reply:
I see what u mean now. Well I was afraid of running the same logic two times for the complex probes when there is no need.

For smaller programs I think it would be fine but for complex programs with lots of args I think it's too much. Also,there was a need for a tailcall so it would be 4 programs doing repetitive stuff in first half of them. 

I think the way you did is more correct and better but the exec has 18 arguments (it's potentially lot of work)

WDYT ? Like you said I think those big programs should be an exception to the control plane logic (just like tailcall are kind of an exception because of big programs)

Reply:
So the only logic that should be duplicated is the argument extraction. What i'm thinking is that we can add a map for sending arguments from the control plane into the event programs (working under the assumption the control plane programs will always run first since they are attached first). We could also use this map in the cgroups control plane and reduce their instruction count.

But even if we have to duplicate argument extraction I don't think it's that bad, the sched_ family isn't the worst in terms of overhead as i've already mentioned.

I think some level of separation is important. Maybe a compromise could be to separate by a tailcall (if ordering isn't guaranteed by attach order).